src/conduits/RwaOutputConduit3.sol

```diff
@@ -0,0 +1,379 @@
+        may[usr] = 1;
+        emit Mate(usr);
+    }
+
+    /**
+     * @notice Revokes `usr` push access from this contract.
+     * @param usr The user address.
+     */
+    function hate(address usr) external auth {
+        may[usr] = 0;
+        emit Hate(usr);
+    }
+
+    /**
+     * @notice Whitelist `who` address for `pick`.
+     * @param who The user address.
+     */
+    function kiss(address who) external auth {
+        bud[who] = 1;
+        emit Kiss(who);
+    }
+
+    /**
+     * @notice Remove `who` address from `pick` whitelist.
+     * @param who The user address.
+     */
+    function diss(address who) external auth {
+        if (to == who) to = address(0);
+        bud[who] = 0;
+        emit Diss(who);
+    }
+
+    /*//////////////////////////////////
+               Administration
+    //////////////////////////////////*/
+
+    /**
+     * @notice Updates a contract parameter.
+     * @param what The changed parameter name. `"quitTo"`
+     * @param data The new value of the parameter.
+     */
+    function file(bytes32 what, address data) external auth {
+        if (what == "quitTo") {
+            quitTo = data;
+        } else {
+            revert("RwaOutputConduit3/unrecognised-param");
+        }
+
+        emit File(what, data);
+    }
+
+    /**
+     * @notice Sets `who` address as the recipient.
+     * @param who Recipient address.
+     * @dev `who` address should have been whitelisted using `kiss`.
+     */
+    function pick(address who) external {
+        require(can[msg.sender] == 1, "RwaOutputConduit3/not-operator");
+        require(bud[who] == 1 || who == address(0), "RwaOutputConduit3/not-bud");
+        to = who;
+        emit Pick(who);
+    }
+
+    /*//////////////////////////////////
+               Operations
+    //////////////////////////////////*/
+
+    /**
+     * @notice Method to swap DAI contract balance to GEM through PSM and push it to the recipient address.
+     * @dev `msg.sender` must have been `mate`d and `to` must have been `pick`ed.
+     */
+    function push() external isMate {
+        _doPush(dai.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Swaps the specified amount of DAI into GEM through the PSM and push it to the recipient address.
+     * @dev `msg.sender` must have been `mate`d and `to` must have been `pick`ed.
+     * @param wad DAI amount.
+     */
+    function push(uint256 wad) external isMate {
+        _doPush(wad);
+    }
+
+    /**
+     * @notice Flushes out any DAI balance to `quitTo` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     */
+    function quit() external isMate {
+        _doQuit(dai.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Flushes out the specified amount of DAI to the `quitTo` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     * @param wad DAI amount.
+     */
+    function quit(uint256 wad) external isMate {
```

Comment:
Is possible to expect that this `quit` function is used just a normal way to extract DAI borrowed to a broker? I'm just wondering if we should have two `push` functions and just treat sending DAI the same than in the version 2. The only difference would be that here we are adding an extra method to swap and send.

Reply:
The normal flow of operation is always swapping, that's why it's put into `push()`. This is consistent with the other conduits.

`quit` should only be ever used if `push` fails &ndash; i.e.: the PSM is empty &ndash; (maybe we should document this @nduchak) and the crypto broker is willing to accept DAI. It's a more convoluted process because if we don't have the `quitTo` address beforehand, governance must act to set it afterwards.

> I'm just wondering if we should have two push functions and just treat sending DAI the same than in the version 2

Initially we called this method `pushDai`, however we changed it so it wouldn't be misinterpreted as simply another normal flow.

Reply:
Ok sounds good, if it can not be a normal path good decision to call it differently.
Maybe in the future we need both as normal paths for another `ilk`, I guess we will need a new version of the conduits in that case.

Reply:
`quit` removed in favor of a generic version of `yank`